File: gui/src/app/lib/filter-dsl/scanner.ts
import { isNumeric } from 'rxjs/internal-compatibility';
import { trimChars } from '../util';

export enum Tag {
  EOF = 'EOF',

  LParen = '(',
  RParen = ')',

  And = 'AND',
  Or = 'OR',

  Eq = '=',
  Neq = '!=',
  Like = '~=',
  Gt = '>',
  Ge = '>=',
  Le = '<=',
  Lt = '<',

  Field = 'FIELD',
  Int = 'INT',
  Float = 'FLOAT',
  Bool = 'BOOL',
  String = 'STRING',
  Null = 'NULL',

  Whitespace = 'WHITESPACE',
  Error = 'ERROR',
}

const keywords = {
  'and': Tag.And,
  'or': Tag.Or,
  'true': Tag.Bool,
  'false': Tag.Bool,
  'null': Tag.Null,
};

export class Tok {

  constructor(
    public tag: Tag,
    public lexeme: string,
    public start: number,
    public end: number,
    public error: string = null
  ) {
  }

  trimLexeme(cutset: string): Tok {
    this.lexeme = trimChars(this.lexeme, cutset);
    return this;
  }
}

export function Scan(str: string, cstMode: boolean = false): Tok[] {
  const scanner = new Scanner(str, cstMode);
  let tokens = [];
  while (true) {
    const tok = scanner.next();
    tokens.push(tok);
    if (tok.tag === Tag.EOF) {
      break;
    }
  }
  return tokens;
}

export class Scanner {

  input: string[] = [];
  pos: number = 0;
  offset: number = 0;
  cstMode: boolean = false; //concrete syntax trees require whitespace chars to be tokenized.

  constructor(str: string, cstMode: boolean = false) {
    this.input = str.split('');
    this.cstMode = cstMode;
  }

  next(): Tok {
    if (!this.cstMode) {
      this.skipWhitespace();
    }
    if (this.atEOF()) {
      return this.emit(Tag.EOF);
    }
    const c = this.nextChar();
    switch (c) {
      case '(':
        return this.emit(Tag.LParen);
      case ')':
        return this.emit(Tag.RParen);
      case '=':
        return this.emit(Tag.Eq);
      case '!':
        if (this.matchNextChar('=')) {
          return this.emit(Tag.Neq);
        }
        return this.emitError('expected = after !');
      case '~':
        if (this.matchNextChar('=')) {
          return this.emit(Tag.Like);
        }
        return this.emitError('expected = after ~');
      case '>':
        if (this.matchNextChar('=')) {
          return this.emit(Tag.Ge);
        }
        return this.emit(Tag.Gt);
      case '<':
        if (this.matchNextChar('=')) {
          return this.emit(Tag.Le);
        }
        return this.emit(Tag.Lt);
      case '"':
        return this.scanString();
      default:
        if (this.isWhitespace(c)) {
          return this.emit(Tag.Whitespace);
        }
        if (this.isValidInputChar(c)) {
          return this.scanField();
        }
        if (this.isStartOfNumber(c)) {
          return this.scanNumber();
        }
        return this.emitError(`unknown entity: ${c}`);
    }
  }

  backtrackWhitespace() {
    while (!this.atEOF() && this.isWhitespace(this.input[this.pos-1])) {
      this.pos--;
    }
    this.offset = this.pos;
  }

  private skipWhitespace() {
    while (!this.atEOF() && this.isWhitespace(this.peekChar())) {
      this.nextChar();
    }
    this.offset = this.pos;
  }

  private isWhitespace(char: string): boolean {
    return char.match(/[\s]/) !== null;
  }

  private scanField(): Tok {
    //todo: not sure about IsNumeric here
    while (!this.atEOF() && (this.isValidInputChar(this.peekChar()) || isNumeric(this.peekChar()))) {
      this.nextChar();
    }
    const tok = this.emit(Tag.Field);
    const tag = keywords[tok.lexeme];
    if (tag !== undefined) {
      tok.tag = tag;
    }
    return tok;
  }

  private scanNumber(): Tok {
    let hasDecimal: boolean = false;
    while (!this.atEOF() && (isNumeric(this.peekChar()) || (this.peekChar() == '.' && !hasDecimal))) {
      const c = this.nextChar();
      hasDecimal = hasDecimal || c === '.';
    }
    if (hasDecimal) {
      return this.emit(Tag.Float);
    }
    return this.emit(Tag.Int);
  }

  private scanString(): Tok {
    while (!this.matchNextChar('"')) {
      if (this.atEOF()) {
        return this.emitError('unclosed quote');
      }
      this.nextChar();
    }
    return this.emit(Tag.String);
  }

  private nextChar(): string {
    this.pos++;
    return this.input[this.pos - 1];
  }

  private matchNextChar(c: string): boolean {
    if (this.atEOF() || this.peekChar() != c) {
      return false;
    }
    this.nextChar();
    return true;
  }

  private peekChar(): string {
    return this.input[this.pos];
  }

  private emit(tag: Tag): Tok {
    const lexeme = this.input.slice(this.offset, this.pos).join('');
    const start = this.offset;
    const end = this.pos;
    // advance
    this.offset = this.pos;
    //emit
    return new Tok(tag, lexeme, start, end);
  }

  private emitError(reason: string): Tok {
    const lexeme = this.input.slice(this.offset, this.input.length).join('');
    const start = this.offset;
    const end = this.input.length;
    // advance
    this.offset = this.input.length;
    //emit
    return new Tok(Tag.Error, lexeme, start, end);
  }

  private atEOF(): boolean {
    return this.pos >= this.input.length || this.pos === -1;
  }

  private isValidInputChar(r: string): boolean {
    return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || r == '_'; // does this work in js?
  }

  private isStartOfNumber(c: string): boolean {
    return isNumeric(c) || c == '-';
  }
}

export function tagPrecedence(tag: Tag): number {
  switch (tag) {
    case Tag.And:
      return 2;
    case Tag.Or:
      return 1;
  }
  return 0;
}

File: server/pkg/filter/scanner.go
package filter

import (
	"fmt"
	"strings"
	"unicode"
)

type tag string

const (
	tagEOF tag = "EOF"

	tagLParen tag = "("
	tagRParen tag = ")"

	tagAnd tag = "AND"
	tagOr  tag = "OR"

	tagEq    tag = "="
	tagNeq   tag = "!="
	tagLike  tag = "~="
	tagFuzzy tag = "~"
	tagGt    tag = ">"
	tagGe    tag = ">="
	tagLe    tag = "<="
	tagLt    tag = "<"

	tagField  tag = "FIELD"
	tagInt    tag = "INT"
	tagFloat  tag = "FLOAT"
	tagBool   tag = "BOOL"
	tagString tag = "STRING"
	tagNull   tag = "NULL"
)

var keywords = map[string]tag{
	"and":   tagAnd,
	"or":    tagOr,
	"true":  tagBool,
	"false": tagBool,
	"null":  tagNull,
}

type token struct {
	tag    tag
	lexeme string
}

func (t token) String() string {
	return fmt.Sprintf("{%s: '%s'}", string(t.tag), t.lexeme)
}

func Scan(str string) ([]token, error) {

	scanner := newScanner(str)

	tokens := make([]token, 0)
	for {
		tok, err := scanner.Next()
		if err != nil {
			return nil, err
		}
		tokens = append(tokens, tok)

		if tok.tag == tagEOF {
			break
		}
	}
	return tokens, nil
}

func newScanner(str string) *scanner {
	return &scanner{
		input:  []rune(str),
		pos:    0,
		offset: 0,
	}
}

type scanner struct {
	input  []rune
	pos    int
	offset int
}

// Gets the next token, advancing the scanner.
func (s *scanner) Next() (token, error) {
	return s.next()
}

func (s *scanner) next() (token, error) {
	s.skipWhitespace()

	if s.atEOF() {
		return s.emit(tagEOF), nil
	}

	switch r := s.nextRune(); r {
	case '(':
		return s.emit(tagLParen), nil
	case ')':
		return s.emit(tagRParen), nil
	case '=':
		return s.emit(tagEq), nil
	case '!':
		if s.matchNextRune('=') {
			return s.emit(tagNeq), nil
		}
		return s.error("expected '='")
	case '~':
		if s.matchNextRune('=') {
			return s.emit(tagLike), nil
		}
		return s.emit(tagFuzzy), nil
	case '>':
		if s.matchNextRune('=') {
			return s.emit(tagGe), nil
		}
		return s.emit(tagGt), nil
	case '<':
		if s.matchNextRune('=') {
			return s.emit(tagLe), nil
		}
		return s.emit(tagLt), nil
	case '"':
		return s.scanString()
	default:
		if isValidFieldRune(r) {
			field, err := s.scanField()
			if err != nil {
				return token{}, err
			}
			return field, nil
		}
		if isStartOfNumber(r) {
			return s.scanNumber(), nil
		}
		return s.error("unknown entity")
	}
}

func (s *scanner) nextRune() rune {
	r := s.input[s.pos]
	s.pos++
	return r
}

// matchNextRune will match the next rune of a multi-run tag e.g. >= !=
func (s *scanner) matchNextRune(r rune) bool {
	if s.atEOF() || s.peekRune() != r {
		return false
	}
	s.nextRune()
	return true
}

func (s *scanner) skipWhitespace() {
	for !s.atEOF() && unicode.IsSpace(s.peekRune()) {
		s.nextRune()
	}
	s.offset = s.pos
}

func (s *scanner) scanField() (token, error) {
	for !s.atEOF() && (isValidFieldRune(s.peekRune()) || isNumber(s.peekRune())) {
		s.nextRune()
	}
	tok := s.emit(tagField)
	if tag, ok := keywords[tok.lexeme]; ok {
		tok.tag = tag
	}
	return tok, nil
}

func (s *scanner) scanNumber() token {
	hasDecimal := false
	for !s.atEOF() && (isNumber(s.peekRune()) || (s.peekRune() == '.' && !hasDecimal)) {
		r := s.nextRune()
		hasDecimal = hasDecimal || r == '.'
	}
	if hasDecimal {
		return s.emit(tagFloat)
	}
	return s.emit(tagInt)
}

func (s *scanner) scanString() (token, error) {
	for !s.matchNextRune('"') {
		if s.atEOF() {
			return s.error("unclosed double quote")
		}
		s.nextRune()
	}
	return trimTokenLexeme(s.emit(tagString), `""`), nil
}

func (s *scanner) atEOF() bool {
	return s.pos >= len(s.input)
}

func (s *scanner) peekRune() rune {
	return s.input[s.pos]
}

func (s *scanner) emit(tag tag) token {
	lexeme := string(s.input[s.offset:s.pos])
	s.offset = s.pos
	return token{tag: tag, lexeme: lexeme}
}

func (s *scanner) error(reason string) (token, error) {
	return token{}, fmt.Errorf("failed to scan string at position %d ('%s'): %s", s.pos, string(s.input[s.offset:s.pos]), reason)
}

func isStartOfNumber(r rune) bool {
	return isNumber(r) || r == '-'
}

func isNumber(r rune) bool {
	return r >= '0' && r <= '9'
}

func isValidFieldRune(r rune) bool {
	return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || r == '_'
}

func trimTokenLexeme(t token, trimSet string) token {
	t.lexeme = strings.Trim(t.lexeme, trimSet)
	return t
}
